### PR TITLE
Implement CReLU (Concatenated ReLU)

### PR DESCRIPTION
**Pull request by muupan**
This PR implements a new activation function CReLU (Concatenated ReLU) described in http://arxiv.org/abs/1603.05201.

### CODE REVIEW

File: tests/chainer_tests/functions_tests/activation_tests/test_crelu.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+import chainer
+from chainer import cuda
+from chainer import functions
+from chainer import gradient_check
+from chainer import testing
+from chainer.testing import attr
+from chainer.testing import condition
+
+
+class _TestCReLU(unittest.TestCase):
```
Comment (unnonouno) [inline tests/chainer_tests/functions_tests/activation_tests/test_crelu.py]: Use parameterized test. I mean:
```
@testing.parameterize(
  {'shape': (5, 4), 'axis': 0}, 
  {'shape': (5, 4), 'axis': 1}, 
  ....
)
```
You can access these values as an attribute of the test class like `self.shape`.

File: chainer/functions/activation/crelu.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    """Concatenated Rectified Linear Unit."""
+
+    def __init__(self, axis=1):
+        self.axis = axis
+
+    def check_type_forward(self, in_types):
+        type_check.expect(
+            in_types.size() == 1,
+            in_types[0].dtype == numpy.float32
+        )
```
Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: `in_types[0].ndim > self.axis` is required

File: chainer/functions/activation/crelu.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+        output_shape = list(input_shape)
+        output_shape[self.axis] *= 2
+        return tuple(output_shape)
+
+    def forward(self, x):
+        x, = x
+        xp = cuda.get_array_module(x)
+        y = xp.empty(self.get_output_shape(x.shape), dtype=numpy.float32)
+        y_former, y_latter = xp.split(y, 2, axis=self.axis)
+        zero = utils.force_type(x.dtype, 0)
```
Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: `zero = x.dtype.type(0)`?

File: chainer/functions/activation/crelu.py (hunk 2 above)
Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: Do you want to support negative axis? If so check both cases. If not please check axis in the initializer.

File: chainer/functions/activation/crelu.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+from chainer import utils
+from chainer.utils import type_check
+
+
+class CReLU(function.Function):
+
+    """Concatenated Rectified Linear Unit."""
+
+    def __init__(self, axis=1):
+        self.axis = axis
```
Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: Check if axis is int.

File: chainer/functions/activation/crelu.py (hunk 2 above)
Comment (muupan) [inline chainer/functions/activation/crelu.py]: I'll support negative axis after https://github.com/pfnet/chainer/issues/1282 is fixed.

**Comment by muupan** [comment]
I've made it accept negative axis and fixed the points you mentioned.

File: chainer/functions/activation/crelu.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        if not isinstance(axis, int):
+            raise TypeError('axis must be an integer value')
+        self.axis = axis
+
+    def check_type_forward(self, in_types):
+        type_check.expect(
+            in_types.size() == 1,
+            in_types[0].dtype == numpy.float32,
+            in_types[0].ndim > self.axis,
+            in_types[0].ndim >= - self.axis
```
Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: Remove a redundant space.

**Comment by unnonouno** [comment]
LGTM except a comment.

File: chainer/functions/activation/crelu.py (hunk 5 above)
Comment (muupan) [inline chainer/functions/activation/crelu.py]: You mean `in_types[0].ndim >= - self.axis` -> `in_types[0].ndim >= -self.axis`?

Comment (unnonouno) [inline chainer/functions/activation/crelu.py]: Yes!

**Comment by muupan** [comment]
I fixed it.

**Comment by unnonouno** [comment]
Great! Thank you.

**Pull request merged**